### ISSUE TITLE
[MJAR-39] Maven Sets Manifest Attribute Values to Multi-Line Strings

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-jar-plugin#230]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** opened **[MJAR-39](https://issues.apache.org/jira/browse/MJAR-39?redirect=false)** and commented

    Maven will set a Manifest attribute's value to a multi-line string, which results in an invalid Manifest.
    By default, on its own, Maven adds a Specification-Title attribute to the Manifest, and sets its value to the POM's <description> element.  If this is a multi-line value in the POM, Maven will insert it unchanged: and yield an invalid Manifest.  Maven then can't even load the jar that it just created itself!
    I have filed another bug relating to the fact that it should not insert such an attribute by default: MJAR-38

---

**Affects:** 2.0

**Attachments:**
- [Jar Specification-Title Tester.zip](https://issues.apache.org/jira/secure/attachment/12716695/Jar+Specification-Title+Tester.zip) (_2.79 kB_)
- [MJAR-39.patch](https://issues.apache.org/jira/secure/attachment/12716741/MJAR-39.patch) (_1.60 kB_)
- [Multiline Manifest Value Bug Tester 2.zip](https://issues.apache.org/jira/secure/attachment/12716727/Multiline+Manifest+Value+Bug+Tester+2.zip) (_2.56 kB_)

**Issue Links:**
- [MJAR-38](https://issues.apache.org/jira/browse/MJAR-38) Maven Puts Arbitrary Extension Definition in JAR Manifest by Default.

1 votes, 3 watchers

**Comment by jira-importer** [comment]
**[Jochen Wiedmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jochen.wiedmann@gmail.com)** commented

Kenney is wrong. In other words, Mike's patch (for MJAR-38 and MJAR-39) should still be applied.

**Comment by jira-importer** [comment]
**[Jerome Lacoste](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lacostej)** commented

> Kenney is wrong. In other words, Mike's patch (for MJAR-38 and MJAR-39) should still be applied.

I am not so sure. Aren't multiline attributes ('continuations') allowed in Manifest files? See http://java.sun.com/j2se/1.4.2/docs/guide/jar/jar.html

Have you guys tried to force using the plexus-archiver 1.0-alpha-6? That's where the fix for PLX-185 went in.

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

HI.  I'm the original poster...

Yes, multiline attributes are possible: continuing lines must start with a single space.

But when I posted the bug, Maven was not obeying the rules -- if I remember right, it also even inserted a fully blank line, which is not legal.

Hope that helps.

**Comment by jira-importer** [comment]
**[Jochen Wiedmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jochen.wiedmann@gmail.com)** commented

Noone discusses the possibility of continuation lines, Steve. The question is: What are continuation lines? IMO, they are one very long (> 72 characters) line, which is split into several physical line. Jerome believes, they may contain new lines before splitting them.

Blank lines are legal, btw. The rule for continuation lines says: blank otherchar* lf, where otherchar may very well be a blank.

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

Hi.

Well, simply, lines in the manifest cannot be longer than 72 bytes -- not characters, but UTF-8 encoded bytes.

So, if a given attribute value causes the first line of the name-value pair to exceed that, it needs to be broken across more than one line.  That's all.

To do that, the lines after the first must be preceeded with a single space -- not any whitespace, but only a single space character.  -- Presumably, if the attribute value so happens to contain some kind of white space at this location, only the first space character is stripped away and the remainder of the line is the continuing attribute value.

But then importantly also, the first blank line in the manifest separates the main attributes section from the rest, the per-entry attributes.  So Maven can't put any fully blank lines in the manifest's main section.  And it also can't do that in individual per-entry sections either.

But it will do that.

Does tht make sense?

I realized that I mixed up the posted test case; I posted a few issues at the same time and mixed files up, and this zip is not useful.  I'll see if I can attach a better test case: but if I fail, you can make a simple POM, and set the \<description> attribute literally like this:

\<description>
This

is a problem.
\</description>

Then Maven will scoop this up and place it into the manifest as the "Specification Title", and it will include the blank line and all the rest.  This Jar is then invalid.

The section in the Jar specification titled "Notes on Manifest and Signature Files" explains the line length and breaking long lines.

Thanks!!
- Steven Coco

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

This is an appropriate test.  The previous zip was a mixed up file, uploaded to the wrong issue.  See the simple README in this zip for instructions.

**Comment by jira-importer** [comment]
**[Jerome Lacoste](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lacostej)** commented

> To do that, the lines after the first must be preceeded with a single space &#8211;
> not any whitespace, but only a single space character. &#8211;

The spec says:

value:                          SPACE *otherchar newline *continuation
continuation:              SPACE *otherchar newline
otherchar:                  any UTF-8 character except NUL, CR and LF

otherchar doesn't exclude SPACE and I believe space is a valid UTF-8 character.

So according to my understanding, a continuation might start with more than one space.

What is invalid is an empty line in the wrong place. The fact that every continuation line starts with a space guarantees that a continuation line won't be mistaken into an "fully blank line".

Now it is true that in maven 2.0.4 (which uses plexus-archiver 1.0-alpha-3) there's an issue. Continuations were wrongly identified and the example \<description> you posted would result in a broken Manifest.

So I would like to see whether the problems still appears if you force the use of a later plexus-archiver (at least version 1.0-alpha-6). I haven't had time to look into your latest test case.

**Comment by jira-importer** [comment]
**[Jochen Wiedmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jochen.wiedmann@gmail.com)** commented

Steve, I can confirm that I can reproduce the problem with maven 2.0.4 and your second zip file.

I have checked, and maven 2.0.4 is indeed using plexus-archiver-1.0-alpha-3, which **does** add carriage returns and line feeds to the manifest value, thus producing invalid manifest entries. The plexus-archiver doesn't, as of 1.0-alpha-4.

I have also verified, that version 2.1-SNAPSHOT of the maven-jar-plugin uses version 2.0.1 of the maven-archiver, which in turn uses plexus-archiver-1.0-alpha-4. In other words, the problem can be fixed by adding

    <version>2.1-SNAPSHOT</version>

to the maven-jar-plugins configuration. (Which I have verified successfully.)

I recommend, that this issue be closed, which I cannot do.
